test/adapters/CharmAdapter.js

```diff
@@ -0,0 +1,734 @@
+                }
+              )
+
+        await time.increaseTo(this.expiry + 1);
+
+        await this.marketContract.settle();
+
+        // For getting expiry exerciseProfit
+        this.exerciseProfit = 0;
+
+        try {
+          this.exerciseProfit = (await this.optionViews.getSellOptionCost(this.market, this.collateralAsset == ONE_ADDRESS ? true : false, this.strikeIndex, this.purchaseAmount)).toString();
+        } catch {}
+
+        assert.equal(
+          (
+            await this.adapter.exerciseProfit(
+              this.oTokenAddress,
+              0,
+              this.purchaseAmount
+            )
+          ).toString(),
+          this.exerciseProfit
+        );
+      });
+    });
+
+    describe("#purchase", () => {
+      let snapshotId;
+
+      beforeEach(async () => {
+        snapshotId = await time.takeSnapshot();
+      });
+
+      afterEach(async () => {
+        await time.revertToSnapShot(snapshotId);
+      });
+
+      it("reverts when buying after expiry", async function () {
+        await time.increaseTo(this.expiry + 1);
+
+        await expect(
+          this.adapter.purchase(
+            [
+              this.underlying,
+              this.strikeAsset,
+              this.collateralAsset,
+              this.expiry,
+              this.strikePrice,
+              this.optionType,
+              this.paymentToken,
+            ],
+            this.purchaseAmount,
+            this.maxCost,
+            {
+              from: user,
+              value: this.premium,
+            }
+          )
+        ).to.be.revertedWith("Cannot purchase after expiry");
+      });
+
+      it("reverts when passing unknown underlying", async function () {
+        await expect(
+          this.adapter.purchase(
+            [
+              "0x0000000000000000000000000000000000000001",
+              this.strikeAsset,
+              this.collateralAsset,
+              this.expiry,
+              this.strikePrice,
+              this.optionType,
+              this.paymentToken,
+            ],
+            this.purchaseAmount,
+            this.maxCost,
+            {
+              from: user,
+              value: this.purchaseAmount,
+            }
+          )
+        ).to.be.reverted;
+      });
+
+      it("purchase mints us tokens", async function () {
+        await depositAndApprove(this.donor, user, USDC_ADDRESS, this.adapter.address);
+        await depositAndApprove(this.donor, user, WBTC_ADDRESS, this.adapter.address);
+        const res = await this.adapter.purchase(
+          [
+            this.underlying,
+            this.strikeAsset,
+            this.collateralAsset,
+            this.expiry,
+            this.strikePrice,
+            this.optionType,
+            this.paymentToken,
+          ],
+          this.purchaseAmount,
+          this.maxCost,
+          {
+            from: user,
+            value: this.premium,
+          }
+        );
```

Comment:
Let's have a benchmark of how much gas will be consumed by a `purchase`, this is what we do for RibbonVol

```
const receipt = await res.wait();
assert.isAtMost(receipt.gasUsed, 1400000);
```